Python/remote_debugging.c

```diff
@@ -0,0 +1,939 @@
+            if (0 != read_memory(
+                    handle,
+                    thread_state_addr + debug_offsets.thread_state.native_thread_id,
+                    sizeof(pid_t),
+                    &this_tid))
+            {
+                return -1;
+            }
+
+            if (this_tid == (unsigned long)tid) {
+                break;
+            }
+
+            if (0 != read_memory(
+                    handle,
+                    thread_state_addr + debug_offsets.thread_state.next,
+                    sizeof(void*),
+                    &thread_state_addr))
+            {
+                return -1;
+            }
+        }
+
+        if (thread_state_addr == 0) {
+            PyErr_SetString(
+                PyExc_RuntimeError,
+                "Can't find the specified thread in the remote process");
+            return -1;
+        }
+    } else {
+        if (0 != read_memory(
+                handle,
+                interpreter_state_addr + debug_offsets.interpreter_state.threads_main,
+                sizeof(void*),
+                &thread_state_addr))
+        {
+            return -1;
+        }
+
+        if (thread_state_addr == 0) {
+            PyErr_SetString(
+                PyExc_RuntimeError,
+                "Can't find the main thread in the remote process");
+            return -1;
+        }
+    }
+
+    uintptr_t eval_breaker;
+    if (0 != read_memory(
+            handle,
+            thread_state_addr + debug_offsets.debugger_support.eval_breaker,
+            sizeof(uintptr_t),
+            &eval_breaker))
+    {
+        return -1;
+    }
+
+    eval_breaker |= _PY_EVAL_PLEASE_STOP_BIT;
+
+    // Ensure our path is not too long
+    if (debug_offsets.debugger_support.debugger_script_path_size <= strlen(debugger_script_path)) {
+        PyErr_SetString(PyExc_ValueError, "Debugger script path is too long");
+        return -1;
+    }
+
+    uintptr_t debugger_script_path_addr = (uintptr_t)(
+        thread_state_addr +
+        debug_offsets.debugger_support.remote_debugger_support +
+        debug_offsets.debugger_support.debugger_script_path);
+    if (0 != write_memory(
+            handle,
+            debugger_script_path_addr,
+            strlen(debugger_script_path) + 1,
+            debugger_script_path))
+    {
+        return -1;
+    }
+
+    int pending_call = 1;
+    uintptr_t debugger_pending_call_addr = (uintptr_t)(
+        thread_state_addr +
+        debug_offsets.debugger_support.remote_debugger_support +
+        debug_offsets.debugger_support.debugger_pending_call);
+    if (0 != write_memory(
+            handle,
+            debugger_pending_call_addr,
+            sizeof(int),
+            &pending_call))
+
+    {
+        return -1;
+    }
+
+    if (0 != write_memory(
+            handle,
+            thread_state_addr + (uintptr_t)debug_offsets.debugger_support.eval_breaker,
+            sizeof(uintptr_t),
+            &eval_breaker))
```

Comment:
We read this value quite a while ago in terms of processing. Any chance we could lose interrupts or something in the remote process? Should we be suspending it while we do all this work? (That would be the typical approach, but maybe it's not generally possible)

Reply:
This is not generally possible without bringing a lot of extra machinery per-architecture and is very dependent on what the caller plants to do. If someone cares they must either send signals, use ptrace or whatever macOs or windows do. It's much much more complex to coordinate that without the context of the caller. The value itself acts like a bunch of different request flags so the worst that can happen is that the eval breaker is visited a bit later for the original pourpose if it fails. I will add some docs about this

Also, less important but relevant still: when benchmarking I have measured that this operation happens so quickly that the chances of any race on this value are practically zero.

Reply:
> We read this value quite a while ago in terms of processing.

It's cheap to move the read closer to the write, though. I'll make that change.